### DISCUSSION TITLE
Doesn't link prediction code need sigmoid function?

### DISCUSSION THREAD

**Discussion by songsong0425** [pyg-team/pytorch_geometric#9306] [Q&A]
Hi, I have a question about the `link_pred` code in the example directory.
When I calculate the probability between nodes, my code returns very large value and need to apply the sigmoid function.
```python
raw_probs = (z[i][nonzero] @ z[i][nonzero].t()) #.sigmoid()

#tensor([[ 486.0474,  167.8329,  327.6024,  ...,   -5.8929,  -38.0647,
#        -215.4613],
#      [ 167.8329,   65.0724,  117.2785,  ...,   -7.8170,  -16.1703,
#          -78.1425],
#        [ 327.6024,  117.2785,  230.3377,  ...,  -13.7387,  -27.1952,
#         -149.6579],
#        ...,
#        [  -5.8929,   -7.8170,  -13.7387,  ...,   41.8735,   14.3932,
#           30.8307],
#        [ -38.0647,  -16.1703,  -27.1952,  ...,   14.3932,   36.4806,
#           32.7753],
#        [-215.4613,  -78.1425, -149.6579,  ...,   30.8307,   32.7753,
#          139.2402]], device='cuda:0')
```

But in the `link_pred` [code](https://github.com/pyg-team/pytorch_geometric/blob/61c47ee404f8e26b3a1cd0db56448b6254920d0e/examples/link_pred.py), there was no sigmoid function for calculating the probabilities.

Can anyone explain if it is a normal phenomenon returning a very large range of values after calculating the edge probability?

**Comment by EtherealClarity** [comment]
Maybe you can check to see if the representation of the model output is using a function like log_softmax() and just output the representation, because this causes the overall value of the representation to become larger.

**Reply by songsong0425** [reply]
Thank you for your prompt reply! Can I ask more detailed explanation? I just used `GATConv` and added original features like skip connection. (Almost similar to this [code](https://github.com/pyg-team/pytorch_geometric/blob/61c47ee404f8e26b3a1cd0db56448b6254920d0e/examples/ppi.py))
I'm sorry for bothering you, but if I couldn't find a function like `log_softmax()`, where should I check? 

[Update]
I applied the `F.normalize` to the input feature and the range of values decreased, but it still returned too many edges whose probability is more 0.99..

**Comment by EtherealClarity** [comment]
> Thank you for your prompt reply! Can I ask more detailed explanation? I just used `GATConv` and added original features like skip connection. (Almost similar to this [code](https://github.com/pyg-team/pytorch_geometric/blob/61c47ee404f8e26b3a1cd0db56448b6254920d0e/examples/ppi.py)) I'm sorry for bothering you, but if I couldn't find a function like `log_softmax()`, where should I check?
> 
> [Update] I applied the `F.normalize` to the input feature and the range of values decreased, but it still returned too many edges whose probability is more 0.99..

it's a pleasure, in my case, I change the code from 
```
def forward(self, data):
        .......
        return F.log_softmax(x, dim=1)
```
to 
```
def forward(self, data):
        .......
        return x
```
This avoids situations where the value of z is too large and result of dot product is very large when calculating the probability of two points connection. You can check your forward function for similar operations that make the node representation larger

**Reply by songsong0425** [reply]
Thank you for your kind reply! Unfortunately, I didn't use the `F.log_softmax` function as the returned value.. :( Are there any other solutions? Now I'm trying to change the slope of the sigmoid function, but I think it is unnecessary job...
